internal/measurexlite/resolver.go

```diff
@@ -0,0 +1,106 @@
+package measurexlite
+
+//
+// Trusted Recursive Resolver
+//
+
+import (
+	"context"
+	"net"
+
+	"github.com/ooni/probe-cli/v3/internal/model"
+	"github.com/ooni/probe-cli/v3/internal/netxlite"
+)
+
+// NewTrustedRecursiveResolver2 returns a new trusted recursive resolver
+func NewTrustedRecursiveResolver2(logger model.Logger, URL string) model.Resolver {
+	const mozillaURL = "https://mozilla.cloudflare-dns.com/dns-query"
+	if URL == "" {
+		URL = mozillaURL
+	}
+	return &TrustedRecursiveResolver2{
+		Logger:                            logger,
+		URL:                               URL,
+		NewParallelDNSOverHTTPSResolverFn: nil,
+		ResolverSystem:                    netxlite.NewStdlibResolver(logger),
+	}
+}
+
+// TrustedRecursiveResolver2 emulates Firefox's TRR2 mode
+type TrustedRecursiveResolver2 struct {
+	Logger model.Logger
+	URL    string
+
+	// NewParallelDNSOverHTTPSResolverFn is OPTIONAL and can be used to overide
+	// calls to the netxlite.NewParallelDNSOverHTTPSResolver factory
+	NewParallelDNSOverHTTPSResolverFn func(model.Logger, string) model.Resolver
+
+	// ResolverSystem is MANADATORY and is used as a fallback if the DoH resolver fails
+	ResolverSystem model.Resolver
+}
+
+var _ model.Resolver = &TrustedRecursiveResolver2{}
+
+// Address implements model.Resolver.Address
+// we can't return a decisive address here since we don't know if LookupHost
+// succeeds with the DoH resolver or the system resolver
+func (trr *TrustedRecursiveResolver2) Address() string {
+	return ""
+}
+
+// Network implements model.Resolver.Network
+// we can't return a decisive network here since we don't know if LookupHost
+// succeeds with the DoH resolver or the system resolver
+func (trr *TrustedRecursiveResolver2) Network() string {
+	return ""
+}
+
+// CloseIdleConnections implements model.Resolver.CloseIdleConnections
+// we only check the fallback resolver since DoH resolver is generated on the fly
+func (trr TrustedRecursiveResolver2) CloseIdleConnections() {
+	trr.ResolverSystem.CloseIdleConnections()
+}
+
+// LookupHost implements model.Resolver.LookupHost
+func (trr *TrustedRecursiveResolver2) LookupHost(ctx context.Context, hostname string) ([]string, error) {
+	reso := trr.NewParallelDNSOverHTTPSResolverFn(trr.Logger, trr.URL)
+	if reso == nil {
+		reso = netxlite.NewParallelDNSOverHTTPSResolver(trr.Logger, trr.URL)
+	}
+	addrs, err := reso.LookupHost(ctx, hostname)
+	if err != nil {
+		addrs, err = trr.ResolverSystem.LookupHost(ctx, hostname)
+	}
+	return addrs, err
+}
+
+// LookupHTTPS implements model.Resolver.LookupHTTPS
+func (trr *TrustedRecursiveResolver2) LookupHTTPS(ctx context.Context, domain string) (*model.HTTPSSvc, error) {
+	reso := trr.NewParallelDNSOverHTTPSResolverFn(trr.Logger, trr.URL)
+	if reso == nil {
+		reso = netxlite.NewParallelDNSOverHTTPSResolver(trr.Logger, trr.URL)
+	}
+	out, err := reso.LookupHTTPS(ctx, domain)
+	if err != nil {
+		out, err = trr.ResolverSystem.LookupHTTPS(ctx, domain)
```

Comment:
Because the system resolver only implements `LookupHost`, we should consider implementing `model.SimpleResolver` rather than `model.Resolver`.

Reply:
> I like the work you did so far. As I mention to you in private, we probably need to figure out some design details around this functionality, which won't cause dramatic changes, so we can discuss them in parallel with the code review proper. I highlighted a bunch of places in which I think you should tweak/refactor your code for improved robustness.
> 
> Additionally, I've seen quite a large coverage drop (-0.5%) which most likely means that we have added a bit too much code that we're not testing. I would recommend for you to check which are the new functions that we're not testing and add tests.

Thank you for the feedback. I have made some changes based on the review. I also added some more tests but the coverage still seems to be decreased, so we probably need to add more tests.